processor/batchprocessor/batch_processor.go

```diff
@@ -195,6 +196,8 @@ func (b *shard) start() {
 		DONE:
 			for {
 				select {
+				case <-b.processor.shutdownCtx.Done():
```

Comment:
Is this thread-safe?

Reply:
Yes, I believe it is. The code only enters this loop after the `shutdownCtx` field is set, after `shutdownC` is closed.

Reply:
That is not true, you read the `shutdownCtx` variable which is written unprotected. Do you guarantee that shard.start(..) is called only after shutdown somewhere else?

Reply:
ok, so I changed things around again, with channels to propagate shutdown context signaling it's done.